server: "secure" all of the debug endpoint

tbg opened the pull request:
Previously, while `/debug/` was inaccessible for external requests by default,
`/debug/pprof/*` enjoyed no such protection. Cleaned this up by putting the
authentication handler in the top-level `HandlerFunc`, where it affects
everything equally.

```
$ curl http://127.0.0.1:8080/debug/pprof/profile
[works]
$ curl http://macts-2:8080/debug/pprof/profile
not allowed (due to the 'server.remote_debugging.mode' setting)
$ ./cockroach sql --insecure -e "set cluster setting server.remote_debugging.mode = 'any';"
SET
$ curl http://macts-2:8080/debug/pprof/profile
[works]
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16238)
<!-- Reviewable:end -->

tschottdorf commented:
cc @Freeaqingme, thanks for reporting!

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Can you extend `acceptance/debug_remote_test.go` to test this?

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16238#-:-KlV6SS999SjhIXfPM6_:b-m9pf97)*
<!-- Sent from Reviewable.io -->

tamird commented on pkg/server/debug.go:
```diff
@@ -67,37 +67,46 @@ var debugRemote = settings.RegisterValidatedStringSetting(
 func handleDebug(w http.ResponseWriter, r *http.Request) {
 	handler, _ := debugServeMux.Handler(r)
-	handler.ServeHTTP(w, r)
+	authingHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
What's the point of constructing this? Don't you just immediately call it?

tschottdorf replied:
No reason, changed it.

tschottdorf commented:
@petermattis Done.

TFTRs!

tamird commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/acceptance/debug_remote_test.go, line 71 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16238#-KlVKKmrHuENhGDNxssm:-KlVKKmrHuENhGDNxssn:bpzqx0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/882ecd2d6a16703869969dbfcd5d63a7006551f2/pkg/acceptance/debug_remote_test.go#L71)):*
> ```Go
> 			}
> 
> 			for i, url := range []string{"/debug/", "/debug/pprof", "/debug/requests"} {
> ```

sub test?

---

*[pkg/server/debug.go, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16238#-KlUslo7BSt9iCHxqNqq-r1-70:-KlVKNHHQVwW1Gw_cB4s:b-2asoq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e80538534ff556b7b554cb05dca5692ab6e71ac/pkg/server/debug.go#L70)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

No reason, changed it.
</blockquote></details>

sorry, I mean why do you need to do this wrapping at all? ISTM that you can just write:
```
if any, _ := authRequest(r); !any {
	http.Error(w, "not allowed (due to the 'server.remote_debugging.mode' setting)", http.StatusForbidden)
	return
}
handler, _ := debugServeMux.Handler(r)
handler.ServeHTTP(w, r)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16238)*
<!-- Sent from Reviewable.io -->

tbg committed:
server: secure all of the debug endpoint

Previously, while `/debug/` was inaccessible for external requests by default,
`/debug/pprof/*` enjoyed no such protection. Cleaned this up by putting the
authentication handler in the top-level `HandlerFunc`, where it affects
everything equally.

```
$ curl http://127.0.0.1:8080/debug/pprof/profile
[works]
$ curl http://macts-2:8080/debug/pprof/profile
not allowed (due to the 'server.remote_debugging.mode' setting)
$ ./cockroach sql --insecure -e "set cluster setting server.remote_debugging.mode = 'any';"
SET
$ curl http://macts-2:8080/debug/pprof/profile
[works]
```

tschottdorf commented:
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/acceptance/debug_remote_test.go, line 71 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16238#-KlVKKmrHuENhGDNxssm:-KlVMRaPIMLeiJ1bI-IQ:byi2b0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/882ecd2d6a16703869969dbfcd5d63a7006551f2/pkg/acceptance/debug_remote_test.go#L71)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

sub test?
</blockquote></details>

That would be possible, but I don't think it adds any benefit.

---

*[pkg/server/debug.go, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16238#-KlUslo7BSt9iCHxqNqq-r1-70:-KlVMjdhZhuuYxy0fji8:b-jewt11) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e80538534ff556b7b554cb05dca5692ab6e71ac/pkg/server/debug.go#L70)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

sorry, I mean why do you need to do this wrapping at all? ISTM that you can just write:
```
if any, _ := authRequest(r); !any {
	http.Error(w, "not allowed (due to the 'server.remote_debugging.mode' setting)", http.StatusForbidden)
	return
}
handler, _ := debugServeMux.Handler(r)
handler.ServeHTTP(w, r)
```
</blockquote></details>

Oh, good call. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16238)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/acceptance/debug_remote_test.go, line 71 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16238#-KlVKKmrHuENhGDNxssm:-KlVOrxrJFUtrC2rsVQQ:b-ze5kif) ([raw file](https://github.com/cockroachdb/cockroach/blob/882ecd2d6a16703869969dbfcd5d63a7006551f2/pkg/acceptance/debug_remote_test.go#L71)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

That would be possible, but I don't think it adds any benefit.
</blockquote></details>

There's a benefit in setting precedent, and it would produce a better failure message.

---

*[pkg/server/debug.go, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16238#-KlUslo7BSt9iCHxqNqq-r1-70:-KlVOoJoTgrS2dGLeGOE:b9gl0zr) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e80538534ff556b7b554cb05dca5692ab6e71ac/pkg/server/debug.go#L70)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Oh, good call. Done.
</blockquote></details>

random newline now, but whatever.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16238)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Thanks for the test!

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16238#-:-KlVPc63lFihMnguXVWH:b56zuc7)*
<!-- Sent from Reviewable.io -->
